File: src/main/java/io/github/svegon/capi/event/render/ShouldEntityRenderCallback.java
package io.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.render.Camera;
import net.minecraft.client.render.Frustum;
import net.minecraft.entity.Entity;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface ShouldEntityRenderCallback {
    Event<ShouldEntityRenderCallback> EVENT = EventFactory.createArrayBacked(ShouldEntityRenderCallback.class,
            (entity, frustum, camera, callback) -> {}, listeners -> (entity, frustum, camera, callback) -> {
        for (ShouldEntityRenderCallback listener : listeners) {
            listener.shouldRenderEntity(entity, frustum, camera, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
            });

    void shouldRenderEntity(Entity entity, Frustum frustum, Camera camera, CallbackInfoReturnable<Boolean> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/ShouldEntityRenderCallback.kt
package io.github.svegon.capi.event.render

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.render.Camera
import net.minecraft.entity.Entity
import java.util.function.Function

interface ShouldEntityRenderCallback {
    fun shouldRenderEntity(
        entity: Entity?,
        frustum: Frustum?,
        camera: Camera?,
        callback: CallbackInfoReturnable<Boolean?>?
    )

    companion object {
        @JvmField
        val EVENT: Event<ShouldEntityRenderCallback?> = EventFactory.createArrayBacked<ShouldEntityRenderCallback?>(
            ShouldEntityRenderCallback::class.java,
            ShouldEntityRenderCallback { entity: Entity?, frustum: Frustum?, camera: Camera?, callback: CallbackInfoReturnable<Boolean?>? -> },
            Function<Array<ShouldEntityRenderCallback?>, ShouldEntityRenderCallback?> { listeners: Array<ShouldEntityRenderCallback?>? ->
                ShouldEntityRenderCallback { entity: Entity?, frustum: Frustum?, camera: Camera?, callback: CallbackInfoReturnable<Boolean?>? ->
                    for (listener in listeners) {
                        listener.shouldRenderEntity(entity, frustum, camera, callback)

                        if (callback.isCancelled()) {
                            return@ShouldEntityRenderCallback
                        }
                    }
                }
            })
    }
}
